storage/src/main/scala/coop/rchain/storage/LMDBStore.scala

```diff
@@ -0,0 +1,284 @@
+package coop.rchain.storage
+
+import java.nio.ByteBuffer
+import java.nio.file.Path
+import java.security.MessageDigest
+
+import cats.implicits._
+import com.google.protobuf.ByteString
+import coop.rchain.storage.util._
+import org.lmdbjava.DbiFlags.MDB_CREATE
+import org.lmdbjava.{Dbi, Env, Txn}
+import coop.rchain.models.Serialize
+import coop.rchain.models.implicits.rhoInstanceWrapper
+import coop.rchain.storage.LMDBStore.{fromBB, fromBL, fromBS, hashBytes, toBB, toBL, toBS}
+import coop.rchain.storage.datamodels.{BytesList, PsKsBytes, PsKsBytesList}
+
+trait IClearableStore {
+
+  /**
+    *  clears database, used mostly to speedup unit-tests, and implemented by LMDBStore only
+    */
+  def clear(): Unit
+}
+
+class LMDBStore[C, P, A, K] private (env: Env[ByteBuffer],
+                                     _dbKeys: Dbi[ByteBuffer],
+                                     _dbPsKs: Dbi[ByteBuffer],
+                                     _dbAs: Dbi[ByteBuffer],
+                                     _dbJoins: Dbi[ByteBuffer])(implicit sc: Serialize[C],
+                                                                pc: Serialize[P],
+                                                                ac: Serialize[A],
+                                                                kc: Serialize[K],
+                                                                bl: Serialize[BytesList])
+    extends IStore[C, P, A, K]
+    with IClearableStore {
+
+  type H = ByteBuffer
+
+  private[storage] def hashC(packedCs: H): H =
+    hashBytes(packedCs)
+
+  private[storage] def hashC(c: C)(implicit serialize: Serialize[C]): H =
+    hashC(List(c))(serialize)
+
+  private[storage] def hashC(cs: List[C])(implicit serialize: Serialize[C]): H =
+    hashC(toBB(cs)(serialize))
+
+  private[storage] def getKey(txn: T, s: H): List[C] =
+    fromBB[C](Option(_dbKeys.get(txn, s))).getOrElse(List.empty)
+
+  private[storage] def putCs(txn: T, channels: List[C]): Unit =
+    putCsH(txn, channels)
+
+  private[storage] def putCsH(txn: T, channels: List[C]): H = {
+    val packedCs = toBB(channels)
+    val hashCs   = hashC(packedCs)
+    _dbKeys.put(txn, hashCs, packedCs)
+    hashCs
+  }
+
+  type T = Txn[ByteBuffer]
+
+  def createTxnRead(): T = env.txnWrite
+
+  def createTxnWrite(): T = env.txnWrite
+
+  def withTxn[R](txn: T)(f: T => R): R =
+    try {
+      val ret: R = f(txn)
+      txn.commit()
+      ret
+    } catch {
+      case ex: Throwable =>
+        txn.abort()
+        throw ex
+    } finally {
+      txn.close()
+    }
+
+  def putA(txn: T, channels: List[C], a: A): Unit = {
+    val hashCs = putCsH(txn, channels)
+    val as     = fromBB[A](Option(_dbAs.get(txn, hashCs))).getOrElse(List.empty[A])
```

Comment:
I prefer:    

```scala
val as = Option(_dbAs.get(txn, hashCs)).map(fromByteBuffer[A]).getOrElse(List.empty[A])
```

See below for a redefinition of `fromBB` to `fromByteBuffer`.